lib/spack/spack/repository.py

```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
         self._all_package_names = None
         self._provider_index = None
 
-        # If repo_dirs is empty, just use the configuration
-        if not repo_dirs:
-            import spack.config
-            repo_dirs = spack.config.get_config('repos')
-            if not repo_dirs:
-                raise NoRepoConfiguredError(
-                    "Spack configuration contains no package repositories.")
-
         # Add each repo to this path.
         for root in repo_dirs:
-            try:
-                repo = Repo(root, self.super_namespace)
-                self.put_last(repo)
-            except RepoError as e:
-                tty.warn("Failed to initialize repository at '%s'." % root,
-                         e.message,
-                         "To remove the bad repository, run this command:",
-                         "    spack repo rm %s" % root)
+            self.append_from_path(root)
+
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```

Comment:
Ok, got the message: *goose-typing* not welcome here. :smile: I'll close this PR as soon as I extract the fix. Anyhow:

> If I want this to behave correctly as a MutableSequence then I have to be familiar with its implementation.

Not True. You have to be familiar with its contract, i.e. you need to know the semantic of the abstract functions and the other parts will follow. Or, said otherwise, this is as true as saying that you need to know the details of Tim sort every time you pass a `key` to `sorted`. I personally don't.